### PR TITLE
Dkryvchuk/284 fingerprints to df

### PR DESCRIPTION
**Pull request by dionissqq**

### CODE REVIEW

**Comment by skalkin** [comment]
Ouch, look like we have not merged it yet :( Let's find a way to proceed

File: libraries/utils/src/bit-array.ts [hunk 1]
```diff
@@ -364,6 +364,24 @@ export default class BitArray {
     this.incrementVersion(notify);
   }
 
+  getRange(from: number, to: number): BitArray {
+    this.assureInRange(from, 0, this._length - 1, 'from');
```
Comment (LaykDimon) [inline libraries/utils/src/bit-array.ts:368]: get rid of magic string

File: packages/Chem/src/chem-searches.ts [hunk 2]
```diff
@@ -12,11 +12,16 @@ import BitArray from '@datagrok-libraries/utils/src/bit-array';
 import {tanimotoSimilarity} from '@datagrok-libraries/utils/src/similarity-metrics';
 import {assure} from '@datagrok-libraries/utils/src/test';
 
-function _chemFindSimilar(molStringsColumn: DG.Column,
-      queryMolString: string, settings: { [name: string]: any }): DG.DataFrame {
+const enum FING_COL_TAGS{
+  invalidatedForVersion = '.invalideted.for.version',
+}
+
+let LastColumnInvalidated: string = '';
```
Comment (LaykDimon) [inline packages/Chem/src/chem-searches.ts:19]: why pascal case here?

File: packages/Chem/src/chem-searches.ts [hunk 3]
```diff
@@ -46,145 +51,169 @@ function _chemFindSimilar(molStringsColumn: DG.Column,
...
-class CacheParams {
-  cachedForCol : DG.Column | null = null;
-  cachedForColVersion: number | null = null;
-  column: DG.Column | null = null;
-  query: string | null = null;
-  moleculesWereIndexed: boolean | null = false;
-  morganFingerprintsWereIndexed: boolean | null = false;
-  morganFingerprints: BitArray[] | null = null;
+function colInvalidated(col: DG.Column): Boolean {
+  return (LastColumnInvalidated == col.name &&
```
Comment (LaykDimon) [inline packages/Chem/src/chem-searches.ts:63]: === preferable instead of ==

File: packages/Chem/src/chem-searches.ts [hunk 4]
```diff
@@ -46,145 +51,169 @@ function _chemFindSimilar(molStringsColumn: DG.Column,
...
-    }
-    if (includeFingerprints) {
-      if (sameColumnAndVersion() && !_chemCache.morganFingerprintsWereIndexed) {
-        await (await getRdKitService()).initMorganFingerprints();
-        _chemCache.morganFingerprintsWereIndexed = true;
-        _chemCache.morganFingerprints = await (await getRdKitService()).getMorganFingerprints();
-        console.log(`Morgan fingerprints for a dataset ${_chemCache.cachedForCol?.name} invalidated`);
+      if (needsUpdate) {
+        // This seems to be the only way to trigger re-calculation of categories
+        molCol.compact();
```
Comment (LaykDimon) [inline packages/Chem/src/chem-searches.ts:85]: no need in { }

File: packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts [hunk 5]
```diff
@@ -45,60 +43,79 @@ export class RdKitServiceWorkerSubstructure extends RdKitServiceWorkerSimilarity
...
+            }
+            if (createdMol) { // check the qmol is proper
+              const match = mol!.get_substruct_match(queryMol!);
+              if (match === '{}') {
+                queryMol!.delete();
+                queryMol = mol;
+              }
+            } // else, this looks to be a real SMARTS
+          } else { // failover to queryMolBlockFailover
+            queryMol = this.getMol(queryMolBlockFailover);
```
Comment (LaykDimon) [inline packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts:97]: no need in { }

File: packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts [hunk 6]
```diff
@@ -45,60 +43,79 @@ export class RdKitServiceWorkerSubstructure extends RdKitServiceWorkerSimilarity
...
+            let createdMol = false;
+            try {
+              mol = this.getMol(queryMolString);
+              createdMol = mol.is_valid();
+            } catch (e) {
+              mol?.delete();
+            }
+            if (createdMol) { // check the qmol is proper
+              const match = mol!.get_substruct_match(queryMol!);
+              if (match === '{}') {
```
Comment (LaykDimon) [inline packages/Chem/src/rdkit-service/rdkit-service-worker-substructure.ts:91]: can we put '{}' into const?
you use it 3 times as I see

**Pull request closed without merging**